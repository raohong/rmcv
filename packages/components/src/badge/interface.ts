import type { SystemStyledComponentProps } from '@rmc-vant/system';
import type {
  ComponentStyleOverrides,
  ComponentThemeConfig,
  JSXIntrinsicElementProps,
} from '../types';
import type { BadgeName } from './classNames';

export type BadgePosition =
  | 'top-left'
  | 'bottom-left'
  | 'top-right'
  | 'bottom-right';

export type BadgeOffsetValue = string | number;

type BadgeBaseProps = {
  /**
   * @description 徽标内容
   */
  content?: React.ReactNode;
  /**
   * @description 徽标 children
   */
  children?: React.ReactNode;
  /**
   * @description 是否是园点
   */
  dot?: boolean;
  /**
   * @description 徽标 className
   */
  className?: string;
  /**
   * @description 徽标 style
   */
  style?: React.CSSProperties;
  /**
   * @description 徽标内容为数字时，最大值
   * @default 99
   */
  max?: number;
  /**
   * @description 当 content 是 0 时，是否显示
   * @default true
   */
  showZero?: boolean;
  /**
   * @description 徽标位置，可选值为 top-left bottom-left bottom-right
   * @default top-right
   */
  position?: BadgePosition;
  /**
   * @description 设置徽标的偏移量，数组的两项分别对应水平和垂直方向的偏移量
   */
  offset?: [BadgeOffsetValue, BadgeOffsetValue] | BadgeOffsetValue;

  classNames?: Partial<Record<BadgeNSlot, string>>;
};

export type BadgeProps = JSXIntrinsicElementProps<BadgeBaseProps> &
  SystemStyledComponentProps;

export type BadgeComponentState = {
  dot: boolean;
  position: BadgePosition;
  fixed: boolean;
  color?: string;
};

export type BadgeNSlot = 'root' | 'wrapper' | 'number' | 'numberWrapper';

export type BadgeSlot =
  | BadgeNSlot
  | 'dot'
  | 'positionTopLeft'
  | 'positionTopRight'
  | 'positionBottomLeft'
  | 'positionBottomRight';

export type BadgeStyleOverrides = ComponentStyleOverrides<BadgeSlot>;

export type BadgeThemeConfig = ComponentThemeConfig<
  typeof BadgeName,
  BadgeProps,
  BadgeStyleOverrides
>;

export type NumberSign = 0 | 1 | -1;
export type NumberDir = 1 | -1;

export type NumberScrollerProps = {
  value: number;
  sign: NumberSign;
  dir: NumberDir;
};

export type NumberScrollerState = {
  diff: number[];
  shouldAnimate: boolean;
  value: number | null;
};
