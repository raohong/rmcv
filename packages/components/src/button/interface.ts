import { SystemStyledComponentProps, SystemSxInterpolation } from '@rmc-vant/system';
import type { RefAttributes } from 'react';
import type { LoadingProps } from '../loading';
import type {
  ComponentStyleOverrides,
  ComponentThemeConfig,
  JSXIntrinsicElementProps,
} from '../types';
import type { ButtonName } from './classNames';

export type ButtonSize = 'large' | 'normal' | 'small' | 'mini';

export type ButtonType = 'default' | 'primary' | 'success' | 'warning' | 'danger';

export type ButtonShape = 'round' | 'square' | 'default';

export type NativeButtonHTMLType = 'button' | 'submit' | 'reset';

export type ButtonVariant = 'outlined' | 'contained';

export type ButtonNSlot = 'root' | 'icon' | 'loadingIcon';

export type ButtonSlot =
  | ButtonNSlot
  | 'loading'
  | 'disabled'
  | 'outlined'
  | 'outlinedSizeMini'
  | 'outlinedSizeSmall'
  | 'outlinedSizeNormal'
  | 'outlinedSizeLarge'
  | 'outlinedDefault'
  | 'outlinedPrimary'
  | 'outlinedSuccess'
  | 'outlinedWarning'
  | 'outlinedDanger'
  | 'outlinedShapeRound'
  | 'outlinedShapeSquare'
  | 'contained'
  | 'containedSizeMini'
  | 'containedSizeSmall'
  | 'containedSizeNormal'
  | 'containedSizeLarge'
  | 'containedDefault'
  | 'containedPrimary'
  | 'containedSuccess'
  | 'containedWarning'
  | 'containedDanger'
  | 'containedShapeRound'
  | 'containedShapeSquare'
  | 'block';

type BaseButtonProps = {
  /**
   * @description 按钮类型
   * @default default
   */
  type?: ButtonType;
  /**
   * @description 按钮尺寸
   * @default normal
   */
  size?: ButtonSize;
  /**
   * @description 自定义颜色
   */
  color?: string;
  /**
   * @description 是否是块级元素
   */
  block?: boolean;
  variant?: ButtonVariant;
  /**
   * @description 按钮形状
   */
  shape?: ButtonShape;
  /**
   * @description 按钮 disabled
   */
  disabled?: boolean;
  /**
   * @description 按钮 icon
   */
  icon?: React.ReactNode;
  /**
   * @description 是否 loading
   */
  loading?: boolean;
  /**
   * @description border 是否是高清
   */
  hairline?: boolean;
  /**
   * @description 是否显示 border
   * @default true
   */
  border?: boolean;
  /**
   * @description loading 图标类型
   */
  loadingType?: LoadingProps['type'];
  /**
   * @description loading 图标大小
   */
  loadingSize?: LoadingProps['size'];
  classNames?: Partial<Record<ButtonNSlot, string>>;
};

type AnchorButtonBaseProps = {
  /**
   * @description 按钮 html 链接
   */
  href?: string;
  target?: string;
} & BaseButtonProps;

export type ButtonComponentState = {
  size: ButtonSize;
  shape: ButtonShape;
  type: ButtonType;
  variant: ButtonVariant;
  disabled: boolean;
  loading: boolean;
  border: boolean;
  block: boolean;
  emptyContent: boolean;
  hairline: boolean;
  color?: string;
  colorIsGradient?: boolean;
};

export type AnchorButtonProps = JSXIntrinsicElementProps<AnchorButtonBaseProps, 'a'>;

type NativeButtonBaseProps = {
  /**
   * 按钮的原始 type
   */
  htmlType?: NativeButtonHTMLType;
} & BaseButtonProps;

export type NativeButtonProps = JSXIntrinsicElementProps<
  NativeButtonBaseProps,
  'button'
>;

export interface WithButton {
  (
    props: { href: string } & Omit<AnchorButtonProps, 'href' | 'htmlType'> &
      RefAttributes<HTMLAnchorElement> &
      SystemStyledComponentProps,
  ): JSX.Element;
  (
    props: { target: string } & Omit<AnchorButtonProps, 'target' | 'htmlType'> &
      RefAttributes<HTMLAnchorElement> &
      SystemStyledComponentProps,
  ): JSX.Element;
  (
    props: Omit<NativeButtonProps, 'href' | 'target'> &
      RefAttributes<HTMLButtonElement> &
      SystemStyledComponentProps,
  ): JSX.Element;
}

export type ButtonProps = AnchorButtonProps &
  NativeButtonProps &
  SystemStyledComponentProps & { activeStyle?: SystemSxInterpolation };

export type ButtonStyleOverrides = ComponentStyleOverrides<
  ButtonComponentState,
  ButtonSlot
>;

export type ButtonThemeConfig = ComponentThemeConfig<
  typeof ButtonName,
  ButtonProps,
  ButtonStyleOverrides
>;
