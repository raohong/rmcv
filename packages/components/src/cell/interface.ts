import { SystemStyledComponentProps } from '@rmc-vant/system';
import React from 'react';
import {
  ComponentStyleOverrides,
  ComponentThemeConfig,
  IntrinsicElementsKeys,
  JSXIntrinsicElementProps,
} from '../types';
import type { CellGroupName, CellName } from './classNames';

export type CellArrowDirection = 'left' | 'up' | 'down' | 'right';

export type CellSize = 'normal' | 'large';

export type CellContextState = {
  bordered?: boolean;
  size?: CellSize;
};

type CellBaseProps = {
  /**
   * @description 左侧标题
   */
  title?: React.ReactNode;
  /**
   * @description 右侧内容
   */
  value?: React.ReactNode;
  /**
   * @description 标题下方的描述信息
   */
  label?: React.ReactNode;
  /**
   * @description 左侧图标或图片链接
   */
  icon?: React.ReactNode;
  /**
   * @description 右侧图标
   */
  rightIcon?: React.ReactNode;
  /**
   * @description 是否显示内边框
   */
  border?: boolean;
  /**
   * @description 单元格大小，可选值为 large
   * @default normal
   */
  size?: CellSize;
  /**
   * @description 是否开启点击反馈
   */
  clickable?: boolean;
  /**
   * @description 是否展示右侧箭头并开启点击反馈
   */
  isLink?: boolean;
  /**
   * @description 是否使内容垂直居中
   */
  center?: boolean;
  /**
   * @description 箭头方向
   * @default right
   */
  arrowDirection?: CellArrowDirection;
  component?: IntrinsicElementsKeys;
  classNames?: Partial<Record<CellNSlot, string>>;
};

export type CellProps = JSXIntrinsicElementProps<CellBaseProps> &
  SystemStyledComponentProps;

export type CellNSlot =
  | 'root'
  | 'label'
  | 'value'
  | 'title'
  | 'arrow'
  | 'icon'
  | 'rightIcon';
export type CellSlot = CellNSlot | 'sizeLarge' | 'sizeNormal';

export type CellComponentState = {
  border: boolean;
  size: CellSize;
  clickable: boolean;
  center: boolean;
  arrowDirection: CellArrowDirection;
};

type CellStyleOverrides = ComponentStyleOverrides<CellComponentState, CellSlot>;

export type CellThemeConfig = ComponentThemeConfig<
  typeof CellName,
  CellProps,
  CellStyleOverrides
>;

type CellGroupBaseProps = {
  /**
   * @description 分组标题
   */
  title?: React.ReactNode;
  /**
   * @description 是否显示外边框
   * @default true
   */
  border?: boolean;
  /**
   * @description 是否展示为圆角卡片风格
   */
  inset?: boolean;
  /**
   * @description 设置 Cell 的大小
   */
  size?: CellSize;
  component?: IntrinsicElementsKeys;
  classNames?: Partial<Record<CellGroupNSlot, string>>;
};

export type CellGroupNSlot = 'root' | 'title' | 'content';
export type CellGroupSlot = CellGroupNSlot | 'border' | 'inset';

export type CellGroupComponentState = {
  border: boolean;
  inset: boolean;
  size: CellSize;
};

type CellGroupStyleOverrides = ComponentStyleOverrides<
  CellGroupComponentState,
  CellGroupSlot
>;

export type CellGroupProps = JSXIntrinsicElementProps<CellGroupBaseProps> &
  SystemStyledComponentProps;

export type CellGroupThemeConfig = ComponentThemeConfig<
  typeof CellGroupName,
  CellGroupProps,
  CellGroupStyleOverrides
>;
