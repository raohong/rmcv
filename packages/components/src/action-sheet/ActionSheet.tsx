import { useControllableValue } from '@rmc-vant/hooks';
import { composeProps, isEmpty } from '@rmc-vant/utils';
import clsx from 'clsx';
import React, { useMemo } from 'react';
import { elementBackgroundHapticFeedback } from '../_styles';
import { getDataOrAriaProps } from '../_utils';
import { useThemeProps } from '../config-provider';
import {
  ActionSheetName,
  actionSheetClassNames,
  composeActionSheetSlotClassNames,
} from './classNames';
import type {
  ActionSheetAction,
  ActionSheetComponentState,
  ActionSheetProps,
} from './interface';
import {
  ActionSheetCancelButton,
  ActionSheetCloseIconSx,
  ActionSheetContent,
  ActionSheetDescription,
  ActionSheetItem,
  ActionSheetItemSubname,
  ActionSheetLoadingIcon,
  ActionSheetRoot,
  ActionSheetTitle,
} from './styles';

const elemFeedback = elementBackgroundHapticFeedback();

const ActionSheet: React.FC<ActionSheetProps> = (props) => {
  const {
    onCancel,
    onClose,
    onSelect,
    closeOnClickAction,
    cancelText,
    onOverlayClick,
    overlayClosable,
    closeIcon,
    title,
    description,
    actions,
    afterClose,
    onBeforeClose,
    className,
    content,
    children,
    classNames,
    closable,
    closeOnPopState = true,
    lazyRender = true,
    safeArea = true,
    overlay = true,
    lockScroll = true,
    round = true,
    ...rest
  } = useThemeProps(ActionSheetName, props);
  const [open, setOpen] = useControllableValue(rest, {
    trigger: 'onOpenChange',
    defaultValue: false,
    valuePropName: 'open',
    defaultValuePropName: 'defaultOpen',
  });

  const componentState: ActionSheetComponentState = useMemo(() => ({}), []);
  const slotClassNames = composeActionSheetSlotClassNames(
    componentState,
    classNames,
  );

  const handleClose = async () => {
    onClose?.();
    setOpen(false);
  };

  const handleActionClick = async (action: ActionSheetAction, index: number) => {
    onSelect?.(action, index);
    action.callback?.(action);

    if (closeOnClickAction) {
      const result = await onBeforeClose?.(action);
      if (result !== false) {
        handleClose();
      }
    }
  };

  const renderCancelBtn = () => {
    if (!isEmpty(cancelText)) {
      return (
        <>
          <ActionSheetCancelButton
            componentState={componentState}
            className={slotClassNames.cancelButton}
            activeStyle={elemFeedback}
            onClick={() => {
              onCancel?.();
              handleClose();
            }}
          >
            {cancelText}
          </ActionSheetCancelButton>
        </>
      );
    }

    return null;
  };

  const renderActionSheets = () =>
    actions?.map((action, index) => {
      const itemComponentState = {
        ...componentState,
        disabled: !!action.disabled,
        loading: !!action.loading,
        color: action.color,
      };

      return (
        <ActionSheetItem
          // eslint-disable-next-line react/no-array-index-key
          key={index}
          activeStyle={elemFeedback}
          disabled={action.disabled}
          className={clsx(
            slotClassNames.item,
            action.className,
            action.loading && actionSheetClassNames.itemLoading,
            action.disabled && actionSheetClassNames.itemDisabled,
          )}
          style={{ color: action.color }}
          onClick={() => {
            if (!action.disabled && !action.loading) {
              handleActionClick(action, index);
            }
          }}
          componentState={itemComponentState}
        >
          {action.loading ? (
            <ActionSheetLoadingIcon
              componentState={componentState}
              className={slotClassNames.loadingIcon}
            />
          ) : (
            <>
              <span>{action.name}</span>
              {!isEmpty(action.subname) && (
                <ActionSheetItemSubname
                  componentState={itemComponentState}
                  className={slotClassNames.itemSubname}
                >
                  {action.subname}
                </ActionSheetItemSubname>
              )}
            </>
          )}
        </ActionSheetItem>
      );
    });

  const hasCustomContent = !isEmpty(content);

  return (
    <>
      <ActionSheetRoot
        {...getDataOrAriaProps(rest)}
        lockScroll={lockScroll}
        overlay={overlay}
        overlayClosable={overlayClosable}
        onOverlayClick={onOverlayClick}
        open={open}
        round={round}
        safeArea={safeArea}
        closeable={closable ?? (!isEmpty(title) || hasCustomContent)}
        onClose={handleClose}
        closeIcon={closeIcon}
        afterClose={afterClose}
        position="bottom"
        componentState={componentState}
        className={clsx(slotClassNames.root, className)}
        closeOnPopstate={closeOnPopState}
        closeIconSx={ActionSheetCloseIconSx}
        classNames={{
          closeIcon: slotClassNames.closeIcon,
        }}
        lazyRender={lazyRender}
      >
        {!isEmpty(title) && (
          <ActionSheetTitle
            componentState={componentState}
            className={slotClassNames.title}
          >
            {title}
          </ActionSheetTitle>
        )}
        {!isEmpty(description) && (
          <ActionSheetDescription
            componentState={componentState}
            className={slotClassNames.description}
          >
            {description}
          </ActionSheetDescription>
        )}
        <ActionSheetContent
          componentState={componentState}
          className={slotClassNames.content}
        >
          {hasCustomContent ? content : renderActionSheets()}
        </ActionSheetContent>
        {!hasCustomContent && renderCancelBtn()}
      </ActionSheetRoot>
      {React.isValidElement(children) &&
        React.cloneElement(
          children,
          composeProps(children.props, {
            onClick: () => setOpen(true),
          }),
        )}
    </>
  );
};

export default ActionSheet;
