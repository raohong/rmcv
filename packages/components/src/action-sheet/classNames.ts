import { generateComponentClassNameUtility } from '@rmc-vant/system';
import {
  ActionSheetComponentState,
  ActionSheetNSlot,
  ActionSheetSlot,
} from './interface';

export const ActionSheetName = 'ActionSheet';

export const {
  actionSheetClassNames,
  composeActionSheetSlotClassNames,
  getActionSheetSlotClassNames,
  getActionSheetClassName,
} = generateComponentClassNameUtility<
  typeof ActionSheetName,
  ActionSheetComponentState,
  ActionSheetSlot,
  ActionSheetNSlot
>(
  ActionSheetName,
  {
    root: true,
    title: true,
    description: true,
    content: true,
    cancelButton: true,
    item: true,
    itemSubname: true,
    loadingIcon: true,
    closeIcon: true,
    itemDisabled: true,
    itemLoading: true,
  },
  () => ({
    root: ['root'],
    title: ['title'],
    closeIcon: ['closeIcon'],
    description: ['description'],
    content: ['content'],
    cancelButton: ['cancelButton'],
    item: ['item'],
    itemSubname: ['itemSubname'],
    loadingIcon: ['loadingIcon'],
  }),
);
