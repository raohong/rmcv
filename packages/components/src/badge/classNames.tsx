import { generateComponentClassNameUtility } from '@rmc-vant/system';
import { camelCase } from '@rmc-vant/utils';
import type { BadgeComponentState, BadgeNSlot, BadgeSlot } from './interface';

export const BadgeName = 'Badge';

export const {
  badgeClassNames,
  getBadgeSlotClassNames,
  composeBadgeSlotClassNames,
} = generateComponentClassNameUtility<
  typeof BadgeName,
  BadgeComponentState,
  BadgeSlot,
  BadgeNSlot
>(
  BadgeName,
  {
    root: true,
    number: true,
    numberWrapper: true,
    wrapper: true,
    dot: true,
    positionBottomLeft: true,
    positionBottomRight: true,
    positionTopLeft: true,
    positionTopRight: true,
  },
  ({ dot, position }) => ({
    root: ['root', camelCase(`position-${position}`)],
    wrapper: ['wrapper', dot && 'dot'],
    number: ['number'],
    numberWrapper: ['numberWrapper'],
  }),
);
