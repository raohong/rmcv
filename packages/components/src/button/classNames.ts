import { generateComponentClassNameUtility } from '@rmc-vant/system';
import { camelCase } from '@rmc-vant/utils';
import { ButtonComponentState, ButtonNSlot, ButtonSlot } from './interface';

export const ButtonName = 'Button';

export const {
  buttonClassNames,
  getButtonSlotClassNames,
  composeButtonSlotClassNames,
} = generateComponentClassNameUtility<
  typeof ButtonName,
  ButtonComponentState,
  ButtonSlot,
  ButtonNSlot
>(
  ButtonName,
  {
    root: true,
    icon: true,
    loading: true,
    disabled: true,
    outlined: true,
    outlinedSizeMini: true,
    outlinedSizeSmall: true,
    outlinedSizeNormal: true,
    outlinedSizeLarge: true,
    outlinedDefault: true,
    outlinedPrimary: true,
    outlinedSuccess: true,
    outlinedWarning: true,
    outlinedDanger: true,
    outlinedShapeRound: true,
    outlinedShapeSquare: true,
    contained: true,
    containedSizeMini: true,
    containedSizeSmall: true,
    containedSizeNormal: true,
    containedSizeLarge: true,
    containedDefault: true,
    containedPrimary: true,
    containedSuccess: true,
    containedWarning: true,
    containedDanger: true,
    containedShapeRound: true,
    containedShapeSquare: true,
    loadingIcon: true,
    block: true,
  },
  ({ variant, size, shape, loading, disabled, type, block }) => ({
    root: [
      'root',
      variant,
      loading && 'loading',
      disabled && 'disabled',
      camelCase(`${variant}-size-${size}`),
      camelCase(`${variant}-${type}`),
      shape !== 'default' && camelCase(`${variant}-shape-${shape}`),
      block && 'block',
    ],
    icon: ['icon'],
    loadingIcon: ['loadingIcon'],
  }),
);
