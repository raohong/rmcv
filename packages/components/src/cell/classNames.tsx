import { generateComponentClassNameUtility } from '@rmc-vant/system';
import { camelCase } from '@rmc-vant/utils';
import type {
  CellComponentState,
  CellGroupComponentState,
  CellGroupNSlot,
  CellGroupSlot,
  CellNSlot,
  CellSlot,
} from './interface';

export const CellName = 'Cell';
export const CellGroupName = 'CellGroup';

export const { cellClassNames, getCellSlotClassNames, composeCellSlotClassNames }
  = generateComponentClassNameUtility<
    typeof CellName,
    CellComponentState,
    CellSlot,
    CellNSlot
  >(
    CellName,
    {
      root: true,
      arrow: true,
      icon: true,
      label: true,
      title: true,
      value: true,
      rightIcon: true,
      sizeLarge: true,
      sizeNormal: true,
    },
    ({ size }) => ({
      root: ['root', camelCase(`size-${size}`)],
      icon: ['icon'],
      title: ['title'],
      value: ['value'],
      label: ['label'],
      rightIcon: ['rightIcon'],
      arrow: ['arrow'],
    }),
  );

export const {
  getCellGroupSlotClassNames,
  composeCellGroupSlotClassNames,
  cellGroupClassNames,
} = generateComponentClassNameUtility<
  typeof CellGroupName,
  CellGroupComponentState,
  CellGroupSlot,
  CellGroupNSlot
>(
  CellGroupName,
  {
    root: true,
    title: true,
    content: true,
    border: true,
    inset: true,
  },
  ({ border, inset }) => ({
    root: ['root', border && 'border', inset && 'inset'],
    title: ['title'],
    content: ['content'],
  }),
);
