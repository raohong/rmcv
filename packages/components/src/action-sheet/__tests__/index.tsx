import { act, fireEvent, render, screen } from '@testing-library/react';
import React, { useState } from 'react';
import ActionSheet, { ActionSheetAction } from '..';
import { actionSheetClassNames } from '../classNames';
import { ActionSheetProps } from '../interface';

const testId = 'action-sheet';

const actions: ActionSheetAction[] = [
  { name: '选项一', color: 'red', className: 'color' },
  { name: '选项二', disabled: true, className: 'disabled' },
  { name: '选项三', loading: true, className: 'loading' },
];

test('render correctly', () => {
  const tree = render(<ActionSheet />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with visible', () => {
  render(<ActionSheet data-testid={testId} open />);

  expect(screen.getByTestId(testId)).toHaveAttribute('aria-hidden', 'false');
});

test('render with title', () => {
  render(
    <ActionSheet data-testid={testId} title={<span data-testid="title" />} open />,
  );

  expect(screen.getByTestId('title')).toBeInTheDocument();
});

test('render with description', () => {
  render(
    <ActionSheet
      data-testid={testId}
      description={<span data-testid="desc" />}
      open
    />,
  );

  expect(screen.getByTestId('desc')).toBeInTheDocument();
  expect(
    screen.getByTestId(testId).querySelector(`.${actionSheetClassNames.closeIcon}`),
  ).toBeInTheDocument();
});

test('render with actions', () => {
  render(<ActionSheet data-testid={testId} actions={actions} open />);

  const container = screen
    .getByTestId(testId)
    .querySelector(`.${actionSheetClassNames.content}`)!;

  expect(container.querySelectorAll(`.${actionSheetClassNames.item}`).length).toBe(
    actions.length,
  );

  expect(container.querySelector('.color')).toBeInTheDocument();
  expect(container.querySelector('.loading')).toBeInTheDocument();
  expect(container.querySelector('.disabled')).toBeInTheDocument();

  expect(container.querySelector('.color')).toHaveStyle({
    color: 'red',
  });
  expect(container.querySelector('.loading')).toHaveClass(
    actionSheetClassNames.itemLoading,
  );
  expect(container.querySelector('.disabled')).toHaveClass(
    actionSheetClassNames.itemDisabled,
  );
});

test('render with onSelect', async () => {
  const onSelect = jest.fn();

  render(
    <ActionSheet data-testid={testId} onSelect={onSelect} actions={actions} open />,
  );

  await act(async () => {
    fireEvent.click(screen.getByTestId(testId).querySelector(`.color`)!);
  });

  expect(onSelect).toHaveBeenCalledWith(actions[0], 0);

  await act(async () => {
    fireEvent.click(screen.getByTestId(testId).querySelector(`.loading`)!);
  });

  expect(onSelect).toHaveBeenCalledTimes(1);

  await act(async () => {
    fireEvent.click(screen.getByTestId(testId).querySelector(`.disabled`)!);
  });

  expect(onSelect).toHaveBeenCalledTimes(1);
});

test('render with closeOnClickAction', async () => {
  const fn = jest.fn();

  const App = () => {
    const [open, set] = useState(true);

    return (
      <ActionSheet
        data-testid={testId}
        actions={actions}
        open={open}
        onOpenChange={set}
        onClose={() => {
          fn();
          set(false);
        }}
        closeOnClickAction
      />
    );
  };

  render(<App />);

  await act(async () => {
    fireEvent.click(screen.getByTestId(testId).querySelector(`.color`)!);
  });

  expect(fn).toHaveBeenCalled();
});

test('render with onCancel', async () => {
  const onCancel = jest.fn();

  render(
    <ActionSheet
      data-testid={testId}
      actions={actions}
      cancelText="cancel"
      onCancel={onCancel}
      open
    />,
  );

  expect(screen.getByText('cancel')).toBeInTheDocument();

  await act(async () => {
    fireEvent.click(screen.getByText('cancel'));
  });

  expect(onCancel).toHaveBeenCalled();
});

test('render with onBeforeClose', async () => {
  const App = (props: ActionSheetProps) => {
    const [open, set] = useState(true);

    return (
      <ActionSheet
        data-testid={testId}
        actions={actions}
        open={open}
        onOpenChange={set}
        closeOnClickAction
        {...props}
      />
    );
  };

  const onBeforeClose = jest.fn(() => false);

  render(<App onBeforeClose={onBeforeClose} />);

  await act(async () => {
    fireEvent.click(screen.getByTestId(testId).querySelector(`.color`)!);
  });

  expect(onBeforeClose).toHaveBeenCalled();

  expect(screen.getByTestId(testId)).toHaveAttribute('aria-hidden', 'false');
});
