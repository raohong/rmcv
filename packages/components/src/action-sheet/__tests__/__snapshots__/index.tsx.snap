// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  position: fixed;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
  touch-action: none;
  z-index: 1;
  background-color: rgba(0, 0, 0, 0.7);
}

<div
    aria-hidden="true"
    class="RVuiOverlay-root RVuiPopup-overlay emotion-0"
    style="display: none; opacity: 0;"
  />
</DocumentFragment>
`;
